### PR TITLE
create model with all necessary associations

### PR DESCRIPTION
**Pull request by marcode95**
In this milestone I created migrations and models with all necessary associations for the friendship functionality.

### CODE REVIEW

**Comment by mimipeshy** [comment]
### STATUS: CHANGES REQUIRED 🔴 ♻️ 
Hello there @marcode95 👋🏾 👋🏾 , great job on this milestone 👏🏾 👏🏾 you are almost done. However, let's make a few changes before marking the project as complete

- [ ] In database.yml do not expose your username and password in production mode, this is a security issue so use the default .yml file provided by Microverse.
* [ ] User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] There is a button to invite the user to friendship on the user's list page and on the single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] Remember to add Unit and Integration tests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

That's all :smile: kindly submit another review when done.

_If you have any questions or something is unclear, comment on this PR and tag my name I will be glad to help_
Regards
Peris Ndanu

**Review by NtwaliHeritier — CHANGES_REQUESTED** [review]
### STATUS: CHANGES REQUESTED🛑🛑🛑
Hello team, great work on the project, I can tell the amount of effort you are putting in.
However let's make some changes
[REQUIRED]: Friends request should be seen also under the name of the invitor.
[REQUIRED]: A user should also see his own post in the timeline

Make the requested changes and ask for review👍
Happy coding💻

**Comment by marcode95** [comment]
Fixed both requested changes. Friend requests now seeable in users list, own posts now seeable in timeline

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -19,6 +19,7 @@
         </div>
       </div>
       <% if current_user %>
+        <%= link_to 'Your Profile',  user_path(current_user), class: 'your-profile-link' %>
         <%= link_to 'Sign out', destroy_user_session_path, method: :delete %>
       <% else %>
         <%= link_to 'Sign in', user_session_path %>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb:25]: - [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  

**The Resources below may come in handy**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

File: app/views/users/_user.html.erb [hunk 2]
```diff
@@ -1,6 +1,16 @@
 <li>
-  Name: <%= user.name %>
-  <span class="profile-link">
-    <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
-  </span>
+    Name: <%= user.name %>
+    <span class="profile-link">
+      <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
+    </span>
+  <% if !Friendship.find_by(user_id: current_user.id, friend_id: user.id) and !Friendship.find_by(user_id: user.id, friend_id: current_user.id) and current_user != user %>
```
Comment (acushlakoncept) [inline app/views/users/_user.html.erb]: There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello Team! Great job! on your level of progress, 👏  

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

- [x] The linters on your PR are not running, kindly fix it. 
- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). Currently, it only shows friend's post.
- [x] Great! I see you have added a few specs, lets also add unit tests for the rest of the models like post, comment, and like. Remember to add integration tests as well. Add unit tests to check if the user's input strings are correctly validated. Also, add association tests.
- [x] I made a few comments on your codebase about removing logic from your views, kindly fix them as well.

#### You can do this 💪 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/users/_user.html.erb (hunk 2 above)
Comment (marcode95) [reply app/views/users/_user.html.erb]: Timeline shows own posts too now, removed most of the logic from the views (your first example was part of the microverse-repo I forked, it wasn't written by me), added specs for likes, comments and posts with all kind of tests

Comment (marcode95) [reply app/views/users/_user.html.erb]: Also all linters should be installed now

File: app/views/users/_user.html.erb [hunk 3]
```diff
@@ -3,4 +3,14 @@
...
     <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
   </span>
-</li>
\ No newline at end of file
+  <% if !check_friendship(@friendship, current_user.id, user.id) and !check_friendship(@friendship, user.id, current_user.id) and current_user != user %>
+    <%= form_with(:url => friendships_path) do |f| %>
+      <%= f.hidden_field :user_id, value: user.id %>
+      <%= f.submit "Invite #{user.name}" %>
+  <% end %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/users/_user.html.erb:11]:
```ruby
  <% if !check_friendship(@friendship, current_user.id, user.id) and !check_friendship(@friendship, user.id, current_user.id) and current_user != user %>
    <%= form_with(:url => friendships_path) do |f| %>
      <%= f.hidden_field :user_id, value: user.id %>
      <%= f.submit "Invite #{user.name}" %>
  <% end %>
```
There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello Team! your progress is commendable, 👍 

- [x]   Posts timeline includes posts written by **logged-in user** and their **friends** ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). Currently, it shows all post whether  friend or not.
- [x] Your linters are still not running @marcode95. Try replacing your `.github/workflows/linters.yml` with [this](https://github.com/acushlakoncept/ror-social-scaffold/blob/master/.github/workflows/linters.yml) or try running the action manually: [Here](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow)

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
...

File: app/views/users/_user.html.erb (hunk 3 above)
Comment (marcode95) [reply app/views/users/_user.html.erb:11]: I fixed the timeline and got the linters to run - I can't find any application logic in _user.html.erb tho. The check_friendship method is defined in helpers and uses application logic from the controller only. Thanks for your detailed reviews!

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Hello Team :wave:
Status: Changes Required :recycle:

Great job so far! :clap:
Let's cover a few details together to make your project better :blush: :arrow_double_down:
REQUIRED:

- [ ] User can accept friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)) or reject friendship request([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). In your app user can just send friendship request.
- [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] Add more tests


That's it, you can submit another code review request after making these changes
With all my respect :pray:

Rindra Josia

**Comment by marcode95** [comment]
Users can already accept and reject friendship on their own profile page, there's also a notification for that showing up on the users index page. The timeline is working fine too. Please be more careful with writing reviews in the future.

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [x] Current user should be able to invite and accept invitations from both **All User** sections, and current **User Profile**. Also have the ability to send an Invitation from other users profile page.


</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] nothing to add


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Comment by marcode95** [comment]
Thanks for your review! It's possible to invite users on the individual show pages now. I checked the requirements for the project and didn't find any indications that one is supposed to have invite and reject buttons in the all users section. If I oversaw something please tell me!

**Review by MahmoudBakr23 — APPROVED** [review]
# **Approved ✔✔**

<hr>

Hello there, great effort and work you've put into this project. I liked the message that shows up on all users page. However, what my teammate earlier in their review meant that you could've assigned the "accept friendship" and "reject friendship" links in the all users page the same way you did in the profile for easier interactions with the app's features. Anyways, congratulations on completing this milestone, and good luck with the next 👍😊
Happy coding 👨‍💻💻

**Pull request merged**